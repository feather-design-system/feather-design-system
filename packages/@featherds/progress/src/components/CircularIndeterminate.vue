<template>
  <div class="indeterminate-container" :class="classes">
    <svg viewBox="0 0 100 100">
      <clipPath id="clip">
        <path
          d="M 50 0 a 50 50 0 0 1 0 100 50 50 0 0 1 1 -100 v 10 a 22 22 0 0 0 0 80 22 22 0 0 0 1 -80"
        />
      </clipPath>
      <foreignObject
        x="0"
        y="0"
        width="100"
        height="100"
        clip-path="url(#clip)"
      >
        <div class="track" />
      </foreignObject>
      <g class="indicator">
        <foreignObject
          x="0"
          y="0"
          width="100"
          height="100"
          clip-path="url(#clip)"
        >
          <div class="gradient" />
        </foreignObject>
        <!-- <circle class="tip" r="5.15" cx="51" cy="5"></circle> -->
        <rect class="tip" x="50" width="3" height="10"></rect>
      </g>
    </svg>
  </div>
</template>

<script lang="ts">
import { computed, defineComponent, toRefs } from "vue";

const props = {
  size: { type: String, required: false, default: "lg" },
};

export default defineComponent({
  props,
  setup(props) {
    const { size } = toRefs(props);

    const classes = computed(() => {
      return {
        sm: size.value === "sm",
        md: size.value === "md",
        lg: size.value === "lg",
      };
    });

    return {
      classes,
      size,
    };
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/themes/variables";

svg {
  max-width: 100px;
}

.feather-progress-container {
  .indicator {
    display: inline-block;
  }
  &.error {
    .indicator {
      display: none;
    }
  }
}
.indeterminate-container {
  margin-bottom: 8px;
  &.sm {
    width: 24px;
    height: 24px;
  }
  &.md {
    width: 32px;
    height: 32px;
  }
  &.lg {
    width: 40px;
    height: 40px;
  }
}
.track {
  width: inherit;
  height: inherit;
  background: var($background);
}
.indicator {
  animation: indeterminate-circular 750ms ease-in-out infinite 250ms;
  transform-origin: center;
}
.gradient {
  width: inherit;
  height: inherit;
  border-radius: 50%;
  background: conic-gradient(
    from 360deg,
    transparent 50%,
    var($secondary) 100%
  );
}
.tip {
  fill: var($secondary);
}

@keyframes indeterminate-circular {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
</style>
