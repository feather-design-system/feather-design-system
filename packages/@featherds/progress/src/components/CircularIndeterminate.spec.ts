import CircularIndeterminate from "./CircularIndeterminate.vue";
import { mount } from "@vue/test-utils";
import axe from "@featherds/utils/test/axe";

import { describe, expect, it } from "vitest";

const getWrapper = (opts: Record<string, unknown>) =>
  mount(CircularIndeterminate, opts);

describe("CircularIndeterminate.vue", () => {
  it("should support large size by default", () => {
    const wrapper = getWrapper({});
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should support various sizes", async () => {
    const wrapper = getWrapper({});

    await wrapper.setProps({ size: "sm" });
    expect(wrapper.element).toMatchSnapshot();

    await wrapper.setProps({ size: "md" });
    expect(wrapper.element).toMatchSnapshot();

    await wrapper.setProps({ size: "lg" });
    expect(wrapper.element).toMatchSnapshot();
  });
});
describe("a11y", () => {
  it("should be accessible", async () => {
    const props = {
      title: "ally",
    };
    const wrapper = getWrapper({ props });
    expect(await axe(wrapper.element)).toHaveNoViolations();
  });
});
