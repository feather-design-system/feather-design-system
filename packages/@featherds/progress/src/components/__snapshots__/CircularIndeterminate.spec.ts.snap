// Vitest Snapshot v1

exports[`CircularIndeterminate.vue > should support large size by default 1`] = `
<div
  class="indeterminate-container lg"
>
  <svg
    viewBox="0 0 100 100"
  >
    <clippath
      id="clip"
    >
      <path
        d="M 50 0 a 50 50 0 0 1 0 100 50 50 0 0 1 1 -100 v 10 a 22 22 0 0 0 0 80 22 22 0 0 0 1 -80"
      />
    </clippath>
    <foreignobject
      clip-path="url(#clip)"
      height="100"
      width="100"
      x="0"
      y="0"
    >
      <div
        class="track"
      />
    </foreignobject>
    <g
      class="indicator"
    >
      <foreignobject
        clip-path="url(#clip)"
        height="100"
        width="100"
        x="0"
        y="0"
      >
        <div
          class="gradient"
        />
      </foreignobject>
      <!-- &lt;circle class=&quot;tip&quot; r=&quot;5.15&quot; cx=&quot;51&quot; cy=&quot;5&quot;&gt;&lt;/circle&gt; -->
      <rect
        class="tip"
        height="10"
        width="3"
        x="50"
      />
    </g>
  </svg>
</div>
`;

exports[`CircularIndeterminate.vue > should support various sizes 1`] = `
<div
  class="indeterminate-container sm"
>
  <svg
    viewBox="0 0 100 100"
  >
    <clippath
      id="clip"
    >
      <path
        d="M 50 0 a 50 50 0 0 1 0 100 50 50 0 0 1 1 -100 v 10 a 22 22 0 0 0 0 80 22 22 0 0 0 1 -80"
      />
    </clippath>
    <foreignobject
      clip-path="url(#clip)"
      height="100"
      width="100"
      x="0"
      y="0"
    >
      <div
        class="track"
      />
    </foreignobject>
    <g
      class="indicator"
    >
      <foreignobject
        clip-path="url(#clip)"
        height="100"
        width="100"
        x="0"
        y="0"
      >
        <div
          class="gradient"
        />
      </foreignobject>
      <!-- &lt;circle class=&quot;tip&quot; r=&quot;5.15&quot; cx=&quot;51&quot; cy=&quot;5&quot;&gt;&lt;/circle&gt; -->
      <rect
        class="tip"
        height="10"
        width="3"
        x="50"
      />
    </g>
  </svg>
</div>
`;

exports[`CircularIndeterminate.vue > should support various sizes 2`] = `
<div
  class="indeterminate-container md"
>
  <svg
    viewBox="0 0 100 100"
  >
    <clippath
      id="clip"
    >
      <path
        d="M 50 0 a 50 50 0 0 1 0 100 50 50 0 0 1 1 -100 v 10 a 22 22 0 0 0 0 80 22 22 0 0 0 1 -80"
      />
    </clippath>
    <foreignobject
      clip-path="url(#clip)"
      height="100"
      width="100"
      x="0"
      y="0"
    >
      <div
        class="track"
      />
    </foreignobject>
    <g
      class="indicator"
    >
      <foreignobject
        clip-path="url(#clip)"
        height="100"
        width="100"
        x="0"
        y="0"
      >
        <div
          class="gradient"
        />
      </foreignobject>
      <!-- &lt;circle class=&quot;tip&quot; r=&quot;5.15&quot; cx=&quot;51&quot; cy=&quot;5&quot;&gt;&lt;/circle&gt; -->
      <rect
        class="tip"
        height="10"
        width="3"
        x="50"
      />
    </g>
  </svg>
</div>
`;

exports[`CircularIndeterminate.vue > should support various sizes 3`] = `
<div
  class="indeterminate-container lg"
>
  <svg
    viewBox="0 0 100 100"
  >
    <clippath
      id="clip"
    >
      <path
        d="M 50 0 a 50 50 0 0 1 0 100 50 50 0 0 1 1 -100 v 10 a 22 22 0 0 0 0 80 22 22 0 0 0 1 -80"
      />
    </clippath>
    <foreignobject
      clip-path="url(#clip)"
      height="100"
      width="100"
      x="0"
      y="0"
    >
      <div
        class="track"
      />
    </foreignobject>
    <g
      class="indicator"
    >
      <foreignobject
        clip-path="url(#clip)"
        height="100"
        width="100"
        x="0"
        y="0"
      >
        <div
          class="gradient"
        />
      </foreignobject>
      <!-- &lt;circle class=&quot;tip&quot; r=&quot;5.15&quot; cx=&quot;51&quot; cy=&quot;5&quot;&gt;&lt;/circle&gt; -->
      <rect
        class="tip"
        height="10"
        width="3"
        x="50"
      />
    </g>
  </svg>
</div>
`;
